File: Array/2780.Minimum_Index_of_a_Valid_Split/C++/2780.Minimum_Index_of_a_Valid_Split.cpp
class Solution {
public:
    int minimumIndex(vector<int> & nums) 
    {
        //====================================//
        // Input type:                        //
        //   - vector<int> ref. (nums vector) //
        // Return type:                       //
        //   - int (result minimun index)     //
        //====================================//

        /*Initialize*/
        ///// Length of nums vector /////
        int len_nums = nums.size();

        ///// Record traversal pointer /////
        int record_trav_ptr = (len_nums - 1);

        ///// Record dictionary, hashmap (i.e. left, right) /////
        unordered_map<int, int> record_left_map, record_right_map;


        /*Hashmap-based loop traversal*/
        ///// Step 1: Looped-traversal for the suffix-prefix index /////
        while (record_trav_ptr >= 0) //Whole
        {
            ((record_right_map[nums[record_trav_ptr]])++); //Keep updating, recording

            (record_trav_ptr--); //Keep updating, traversing

        } //Whole

        ///// Step 2: Looped-traversal for the prefix-suffix index /////
        (record_trav_ptr++); //Update, Traverse

        while (record_trav_ptr < len_nums) //Whole
        {
            int record_num = nums[record_trav_ptr]; //Record num

            int split_size = (record_trav_ptr + 1); //Split-size

            ((record_right_map[record_num])--); //Keep updating, recording
            ((record_left_map[record_num])++); //Keep updating, recording

            ///// Check if the current indexed-value is majority or not /////
            if ((record_right_map[record_num] > ((len_nums - split_size) / 2)) && (record_left_map[record_num] > (split_size / 2)))
            {
                return record_trav_ptr;
            }
            else { ; }

            (record_trav_ptr++); //Keep updating, traversing

        } //Whole

        return (-1);
    }
};

File: Array/2780.Minimum_Index_of_a_Valid_Split/Python/2780.Minimum_Index_of_a_Valid_Split.py
class Solution(object):
    def minimumIndex(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        #================================#
        # Hashmap-based traversal method #
        #================================#

        ############
        #Initialize
        ##### Length of nums vector /////
        len_nums = len(nums)

        ##### Record traversal pointer #####
        record_trav_ptr = (len_nums - 1)

        ##### Record dictionary/hashmap (i.e. left, right) #####
        record_left_dict, record_right_dict = {}, {}


        ##############################
        #Hashmap-based loop traversal
        ##### Step 1: Looped-traversal for the suffix-prefix index #####
        while (record_trav_ptr >= 0):

            ##### Check if the current indexed-value existed or not #####
            if (nums[record_trav_ptr] not in record_right_dict):
                record_right_dict[nums[record_trav_ptr]] = 1 #Keep updating/recording
            else:
                record_right_dict[nums[record_trav_ptr]] += 1 #Keep updating/recording

            record_trav_ptr -= 1 #Keep updating/traversing

        ##### Step 2: Looped-traversal for the prefix-suffix index #####
        record_trav_ptr += 1 #Update/Traverse

        while (record_trav_ptr < len_nums):
            record_num = nums[record_trav_ptr] #Record num

            split_size = (record_trav_ptr + 1) #Split-size

            ##### Check if the current indexed-value existed or not #####
            if (record_num not in record_left_dict):
                record_left_dict[record_num] = 1 #Keep updating/recording
            else:
                record_left_dict[record_num] += 1 #Keep updating/recording

            record_right_dict[record_num] -= 1 #Keep updating/recording

            ##### Check if the current indexed-value is majority or not #####
            if ((record_left_dict[record_num] > (split_size // 2)) and (record_right_dict[record_num] > ((len_nums - split_size) // 2))):
                return record_trav_ptr

            record_trav_ptr += 1 #Keep updating/traversing

        return (-1)
